File: android-native/app/src/main/java/com/snape/flix/data/WatchProviders.kt
package com.snape.flix.data

/**
 * The six "preferred" streaming providers the web detail page surfaces, in the
 * same fixed display order. Mirrors `lib/watch-providers.ts`. [asset] points at a
 * bundled white-able SVG in `assets/providers/`.
 */
enum class WatchProviderKey(
    val label: String,
    val asset: String,
    val tmdbId: Int,
    val brandColor: Long,
    val heroId: Int,
    val heroIsSeries: Boolean,
) {
    NETFLIX("Netflix", "file:///android_asset/providers/netflix.svg", 8, 0xFFE50914, 66732, true),
    MAX("Max", "file:///android_asset/providers/max.svg", 1899, 0xFF8A2BE2, 94997, true),
    PRIMEVIDEO("Prime Video", "file:///android_asset/providers/primevideo.svg", 9, 0xFF00A8E0, 76479, true),
    DISNEYPLUS("Disney+", "file:///android_asset/providers/disneyplus.svg", 337, 0xFF166534, 84958, true),
    PARAMOUNTPLUS("Paramount+", "file:///android_asset/providers/paramountplus.svg", 531, 0xFFCA8A04, 73586, true),
    APPLETV("Apple TV+", "file:///android_asset/providers/appletv.svg", 350, 0xFF6E6E73, 438631, false);

    /** Netflix & Max wordmarks read smaller than the others, like the web sizing. */
    val isWide: Boolean get() = this == NETFLIX || this == MAX
}

private fun normalize(name: String): String =
    name.lowercase().replace("+", "plus").replace(Regex("[^a-z0-9]+"), " ").trim()

private fun keyFor(name: String): WatchProviderKey? {
    val n = normalize(name)
    return when {
        n.contains("netflix") -> WatchProviderKey.NETFLIX
        n.contains("prime video") || n.contains("amazon prime") -> WatchProviderKey.PRIMEVIDEO
        n.contains("paramount") -> WatchProviderKey.PARAMOUNTPLUS
        n.contains("apple tv") -> WatchProviderKey.APPLETV
        n == "max" || n.contains("hbo max") || (n.contains("max") && !n.contains("imax")) -> WatchProviderKey.MAX
        n.contains("disney") -> WatchProviderKey.DISNEYPLUS
        else -> null
    }
}

/** Pick the preferred providers present in one region, in fixed order (Apple TV+ last). */
fun pickProviders(region: TmdbProviderRegion?): List<WatchProviderKey> {
    if (region == null) return emptyList()
    val found = LinkedHashSet<WatchProviderKey>()
    listOfNotNull(region.flatrate, region.free, region.ads, region.rent, region.buy)
        .flatten()
        .forEach { p -> keyFor(p.provider_name)?.let { found.add(it) } }

    val ordered = WatchProviderKey.entries.filter { it in found }.toMutableList()
    if (ordered.size > 1 && ordered.remove(WatchProviderKey.APPLETV)) {
        ordered.add(WatchProviderKey.APPLETV)
    }
    return ordered
}

File: lib/watch-providers.ts
import type { TMDBWatchProvidersRegion, TMDBWatchProvider } from '@/types/tmdb';

export type PreferredProviderKey =
  | 'netflix'
  | 'primevideo'
  | 'paramountplus'
  | 'appletv'
  | 'max'
  | 'disneyplus';

export const PREFERRED_PROVIDERS: ReadonlyArray<{
  key: PreferredProviderKey;
  label: string;
  assetPath: string;
  tmdbId: number;
  brandColor: string;
}> = [
  {
    key: 'netflix',
    label: 'Netflix',
    assetPath: '/providers/netflix.svg',
    tmdbId: 8,
    brandColor: '#E50914',
  },
  {
    key: 'max',
    label: 'Max',
    assetPath: '/providers/max.svg',
    tmdbId: 1899,
    brandColor: '#8A2BE2',
  },
  {
    key: 'primevideo',
    label: 'Prime Video',
    assetPath: '/providers/primevideo.svg',
    tmdbId: 9,
    brandColor: '#00A8E0',
  },
  {
    key: 'disneyplus',
    label: 'Disney+',
    assetPath: '/providers/disneyplus.svg',
    tmdbId: 337,
    brandColor: '#166534',
  },
  {
    key: 'paramountplus',
    label: 'Paramount+',
    assetPath: '/providers/paramountplus.svg',
    tmdbId: 531,
    brandColor: '#CA8A04',
  },
  {
    key: 'appletv',
    label: 'Apple TV+',
    assetPath: '/providers/appletv.svg',
    tmdbId: 350,
    brandColor: '#6E6E73',
  },
] as const;

function normalizeProviderName(name: string): string {
  return name
    .toLowerCase()
    .replace(/\+/g, 'plus')
    .replace(/[^a-z0-9]+/g, ' ')
    .trim();
}

function keyFromProvider(p: TMDBWatchProvider): PreferredProviderKey | null {
  const n = normalizeProviderName(p.provider_name);
  if (n.includes('netflix')) return 'netflix';
  if (n.includes('prime video') || n.includes('amazon prime')) return 'primevideo';
  if (n.includes('paramount')) return 'paramountplus';
  if (n.includes('apple tv')) return 'appletv';
  // TMDB may report "HBO Max" or "Max"
  if (n === 'max' || n.includes('hbo max') || (n.includes('max') && !n.includes('imax')))
    return 'max';
  if (n.includes('disney')) return 'disneyplus';
  return null;
}

export function pickPreferredProviders(
  region: TMDBWatchProvidersRegion | null | undefined,
): PreferredProviderKey[] {
  if (!region) return [];
  const buckets = ['flatrate', 'free', 'ads', 'rent', 'buy'] as const;
  const found = new Set<PreferredProviderKey>();

  for (const bucket of buckets) {
    const list = region[bucket];
    if (!list) continue;
    for (const p of list) {
      const key = keyFromProvider(p);
      if (key) found.add(key);
    }
  }

  // Return in a fixed order so UI is consistent.
  const ordered = PREFERRED_PROVIDERS.filter((p) => found.has(p.key)).map((p) => p.key);

  // If multiple providers exist, always show Apple TV+ at the end.
  if (ordered.length > 1) {
    const i = ordered.indexOf('appletv');
    if (i !== -1) ordered.push(...ordered.splice(i, 1));
  }

  return ordered;
}

export function pickPreferredProvidersWithFallback(
  results: Record<string, TMDBWatchProvidersRegion | undefined> | null | undefined,
  country: string,
): PreferredProviderKey[] {
  if (!results) return [];

  const primary = pickPreferredProviders(results[country]);
  if (primary.length > 0) return primary;

  const us = pickPreferredProviders(results.US);
  if (us.length > 0) return us;

  for (const region of Object.values(results)) {
    const picked = pickPreferredProviders(region);
    if (picked.length > 0) return picked;
  }

  return [];
}

export function preferredProviderMeta(key: PreferredProviderKey): {
  label: string;
  assetPath: string;
} {
  const meta = PREFERRED_PROVIDERS.find((p) => p.key === key);
  if (!meta) return { label: key, assetPath: '' };
  return { label: meta.label, assetPath: meta.assetPath };
}
